core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeImageFilter.java

```diff
@@ -52,71 +46,86 @@ Images filter(List<Image> availableImages, Versions supportedVersions, Image cur
+                image.getPackageVersions().get(key));
     }
 
     private Predicate<Image> validateCloudPlatform(String cloudPlatform) {
-        return image -> image.getImageSetsByProvider().keySet().stream().anyMatch(key -> key.equalsIgnoreCase(cloudPlatform));
+        return image -> {
+            boolean result = image.getImageSetsByProvider().keySet().stream().anyMatch(key -> key.equalsIgnoreCase(cloudPlatform));
+            if (!result) {
+                reason = String.format("There are no image available for %s cloud platform.", cloudPlatform);
+            }
+            return result;
+        };
     }
 
-    private Predicate<Image> validateCfmVersion(Image currentImage) {
-        return image -> permitExtensionUpgrade(currentImage, image, CFM_PACKAGE_KEY);
-    }
-
-    private Predicate<Image> validateCspVersion(Image currentImage) {
-        return image -> permitExtensionUpgrade(currentImage, image, CSP_PACKAGE_KEY);
+    private Predicate<Image> validateOsVersion(Image currentImage) {
+        return image -> {
+            boolean result = isOsVersionsMatch(currentImage, image);
+            setReason(result, "There are no other image with the same OS version.");
+            return result;
+        };
     }
 
-    private boolean permitCmAndSatckUpgrade(Image currentImage, Image image, String key) {
-        return upgradePermissionProvider.permitCmAndSatckUpgrade(currentImage.getPackageVersions().get(STACK_PACKAGE_KEY),
-                image.getPackageVersions().get(key));
+    private boolean isOsVersionsMatch(Image currentImage, Image newImage) {
+        return newImage.getOs().equalsIgnoreCase(currentImage.getOs()) && newImage.getOsType().equalsIgnoreCase(currentImage.getOsType());
     }
 
-    private boolean permitExtensionUpgrade(Image currentImage, Image image, String key) {
-        return upgradePermissionProvider.permitExtensionUpgrade(currentImage.getPackageVersions().get(STACK_PACKAGE_KEY),
-                image.getPackageVersions().get(key));
+    private Predicate<Image> validateSaltVersion(Image currentImage) {
+        return image -> {
+            boolean result = image.getPackageVersions().get(SALT_PACKAGE_KEY).equals(currentImage.getPackageVersions().get(SALT_PACKAGE_KEY));
+            setReason(result, "There are no other image with the same salt version.");
+            return result;
+        };
     }
 
-    private Predicate<Image> validateSaltVersion(Image currentImage) {
-        return image -> image.getPackageVersions().get(SALT_PACKAGE_KEY).equals(currentImage.getPackageVersions().get(SALT_PACKAGE_KEY));
+    private String getReason(List<Image> images) {
+        return images.isEmpty() ? reason : null;
     }
 
-    private Predicate<Image> filterCurrentImage(Image currentImage) {
-        return image -> !image.getUuid().equals(currentImage.getUuid());
+    private void setReason(boolean result, String reasonText) {
+        if (!result) {
+            reason = reasonText;
```

Comment:
Wouldn't it be better to aggregate all the validation errors here?

Reply:
You mean we should collect all the validation errors and return them?

Reply:
Yes

Reply:
We discussed in other channel.